src/core/middlewares/regtoken/handler.go

```diff
@@ -0,0 +1,77 @@
+package regtoken
+
+import (
+	"context"
+	"github.com/docker/distribution/registry/auth"
+	"github.com/goharbor/harbor/src/common/rbac"
+	"github.com/goharbor/harbor/src/common/utils/log"
+	"github.com/goharbor/harbor/src/core/config"
+	"github.com/goharbor/harbor/src/core/middlewares/util"
+	pkg_token "github.com/goharbor/harbor/src/pkg/token"
+	"github.com/goharbor/harbor/src/pkg/token/claims/registry"
+	"net/http"
+	"strings"
+)
+
+// regTokenHandler is responsible for decoding the registry token in the docker pull request header,
+// as harbor adds customized claims action into registry auth token, the middlerware is for decode it and write it into
+// request context, then for other middlerwares in chain to use it to bypass request validation.
+type regTokenHandler struct {
+	next http.Handler
+}
+
+// New ...
+func New(next http.Handler) http.Handler {
+	return &regTokenHandler{
+		next: next,
+	}
+}
+
+// ServeHTTP ...
+func (r *regTokenHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+
+	imgRaw := req.Context().Value(util.ImageInfoCtxKey)
+	if imgRaw == nil || !config.WithClair() {
+		r.next.ServeHTTP(rw, req)
+		return
+	}
+	img, _ := req.Context().Value(util.ImageInfoCtxKey).(util.ImageInfo)
```

Comment:
Code here can be simplified as the image info has already been extracted out.

Reply:
There is a function `ImageInfoFromContext` in `middlewares/util` pkg.

Reply:
updated.